buffer/src/main/java/io/netty/buffer/TestingPooledByteBufAllocator.java

```diff
@@ -0,0 +1,133 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.netty.buffer;
+
+import io.netty.util.ResourceLeakDetector;
+import io.netty.util.ResourceLeakTracker;
+
+import java.io.Closeable;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+
+import static io.netty.util.internal.StringUtil.NEWLINE;
+
+/**
+ * A {@link ByteBufAllocator} for unit testing.  This allocator performs leak tracking
+ * on all allocated buffers.  At the end the test, call {@link #close} to assert
+ * that all buffers have been freed.
+ */
+public class TestingPooledByteBufAllocator extends PooledByteBufAllocator implements Closeable {
+
+    private final TestingResourceLeakDetector leakDetector = new TestingResourceLeakDetector();
+
+    public TestingPooledByteBufAllocator() {
+    }
+
+    public TestingPooledByteBufAllocator(boolean preferDirect) {
+        super(preferDirect);
+    }
+
+    @Override
+    ByteBuf toLeakAwareBufferInternal(ByteBuf buf) {
+        ResourceLeakTracker<ByteBuf> leak = leakDetector.track(buf);
+        if (leak != null) {
+            buf = new AdvancedLeakAwareByteBuf(buf, leak);
+        }
+        return buf;
+    }
+
+    @Override
+    CompositeByteBuf toLeakAwareBufferInternal(CompositeByteBuf buf) {
```

Comment:
Overloading with sub/super is probably not going to work out well.   Consider:

```java
ByteBuf composite = new CompositeByteBuffer(...);
toLeakAwareBufferInternal(composite);
```

That will call the wrong tracker.